pkg/agent/controller/trafficcontrol/controller.go

```diff
@@ -0,0 +1,1020 @@
+		portName = genERSPANPortName(port.ERSPAN)
+	}
+	return portName
+}
+
+// getOrCreateTrafficControlPort ensures that there is an OVS port for the given TrafficControlPort and binds the port
+// to the TrafficControl. The OVS port will be created if the port doesn't exist. It returns the ofPort of the OVS port
+// on success, a boolean indicating whether the port exists, and an error if there is.
+func (c *Controller) getOrCreateTrafficControlPort(port *v1alpha2.TrafficControlPort, portName, tcName string) (uint32, bool, error) {
+	c.ovsPortUpdateMutex.Lock()
+	defer c.ovsPortUpdateMutex.Unlock()
+
+	if itf, ok := c.interfaceStore.GetInterfaceByName(portName); ok {
+		// Bind the port to the TrafficControl.
+		if _, exists := c.portToTCBindings[portName]; !exists {
+			c.portToTCBindings[portName] = &portToTCBinding{
+				ofPort:          uint32(itf.OFPort),
+				trafficControls: sets.NewString(),
+			}
+		}
+		c.portToTCBindings[portName].trafficControls.Insert(tcName)
+		return uint32(itf.OFPort), true, nil
+	}
+
+	var portUUID string
+	var err error
+
+	switch {
+	case port.OVSInternal != nil:
+		portUUID, err = c.createOVSInternalPort(portName)
+	case port.Device != nil:
+		portUUID, err = c.ovsBridgeClient.CreatePort(portName, portName, trafficControlPortExternalIDs)
+	case port.VXLAN != nil:
+		portUUID, err = c.createUDPTunnelPort(portName, ovsconfig.VXLANTunnel, port.VXLAN)
+	case port.GENEVE != nil:
+		portUUID, err = c.createUDPTunnelPort(portName, ovsconfig.GeneveTunnel, port.GENEVE)
+	case port.GRE != nil:
+		portUUID, err = c.createGREPort(portName, port.GRE)
+	case port.ERSPAN != nil:
+		portUUID, err = c.createERSPANPort(portName, port.ERSPAN)
+	}
+
+	if err != nil {
+		return 0, false, err
+	}
+
+	ofPort, err := c.ovsBridgeClient.GetOFPort(portName, false)
+	if err != nil {
+		return 0, false, err
+	}
+
+	itf := interfacestore.NewTrafficControlInterface(portName)
+	itf.OVSPortConfig = &interfacestore.OVSPortConfig{PortUUID: portUUID, OFPort: ofPort}
+	c.interfaceStore.AddInterface(itf)
+	// Create binding for the newly created port.
+	c.portToTCBindings[portName] = &portToTCBinding{
+		ofPort:          uint32(ofPort),
+		trafficControls: sets.NewString(tcName),
+	}
+	return uint32(ofPort), false, nil
+}
+
+// releaseTrafficControlPort releases the port from the TrafficControl and deletes the port if it is no longer used by
+// any TrafficControl. It returns the ofPort on success, a boolean indicating whether the port is deleted, and an error
+// if there is.
+func (c *Controller) releaseTrafficControlPort(portName, tcName string) (uint32, bool, error) {
+	c.ovsPortUpdateMutex.Lock()
+	defer c.ovsPortUpdateMutex.Unlock()
+	portBinding := c.portToTCBindings[portName]
+	ofPort := portBinding.ofPort
+
+	portBinding.trafficControls.Delete(tcName)
+	// If the port is no longer used by any TrafficControl, delete the port.
+	if len(portBinding.trafficControls) == 0 {
+		if itf, ok := c.interfaceStore.GetInterfaceByOFPort(ofPort); ok {
```

Comment:
If you just store `*InterfaceConfig` in `portBinding` instead of `ofPort`, the query can be saved.

Reply:
Done.